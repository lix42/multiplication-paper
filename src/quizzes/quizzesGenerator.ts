export type Num = 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;

const minNum = 2;
const maxNum = 12;

export type Quiz = {
  readonly first: Num;
  readonly second: Num;
};

const isNum = (n?: number): n is Num =>
  n != null && Number.isFinite(n) && n >= minNum && n <= maxNum;

const frequency: { [key in Num]: number } = {
  2: 1,
  3: 1,
  4: 1,
  5: 1,
  6: 2,
  7: 2,
  8: 2,
  9: 2,
  10: 1,
  11: 1,
  12: 1,
};
export const generateAllQuizzes = (): Quiz[] => {
  let result: Quiz[] = [];
  for (let first = minNum; first <= maxNum; first++) {
    for (let second = minNum; second <= first; second++) {
      if (isNum(first) && isNum(second)) {
        for (let i = 0; i < frequency[first] * frequency[second]; i++) {
          result.push({ first, second });
        }
      }
    }
  }
  return result;
};

export const pickQuizzes = (candidates: Quiz[], count: number): Quiz[] => {
  let candidatesLength = candidates.length;
  return candidates.reduce((accu: Quiz[], quiz, index) => {
    const candidateCount = candidatesLength - index;
    const resultLeft = count - accu.length;
    if (Math.random() * candidateCount < resultLeft) {
      accu.push(quiz);
    }
    return accu;
  }, []);
};

export const randomSortQuiz = (quiz: Quiz): Quiz =>
  Math.random() * 2 > 1
    ? quiz
    : {
        first: quiz.second,
        second: quiz.first,
      };
