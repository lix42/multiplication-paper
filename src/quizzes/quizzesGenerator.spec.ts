import { pickQuizzes, Quiz } from "./quizzesGenerator";

describe("pickQuizzes", () => {
  const candidates: Quiz[] = [
    {
      first: 2,
      second: 2,
    },
    {
      first: 3,
      second: 3,
    },
    {
      first: 4,
      second: 4,
    },
  ];
  it("should return an empty array if target count is 0", () => {
    const result = pickQuizzes(candidates, 0);
    expect(result).toEqual([]);
  });
  it("should return random quizzes", () => {
    for (let i = 1; i <= 3; i++) {
      const result = pickQuizzes(candidates, i);
      expect(result).toHaveLength(i);
      const set = new Set(result);
      expect(set.size).toBe(i);
    }
  });
});
