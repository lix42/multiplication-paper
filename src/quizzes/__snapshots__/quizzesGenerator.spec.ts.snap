// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateAllQuizzes should return all test cases 1`] = `
Array [
  Object {
    "first": 2,
    "second": 2,
  },
  Object {
    "first": 3,
    "second": 2,
  },
  Object {
    "first": 3,
    "second": 3,
  },
  Object {
    "first": 4,
    "second": 2,
  },
  Object {
    "first": 4,
    "second": 3,
  },
  Object {
    "first": 4,
    "second": 4,
  },
  Object {
    "first": 5,
    "second": 2,
  },
  Object {
    "first": 5,
    "second": 3,
  },
  Object {
    "first": 5,
    "second": 4,
  },
  Object {
    "first": 5,
    "second": 5,
  },
  Object {
    "first": 6,
    "second": 2,
  },
  Object {
    "first": 6,
    "second": 2,
  },
  Object {
    "first": 6,
    "second": 3,
  },
  Object {
    "first": 6,
    "second": 3,
  },
  Object {
    "first": 6,
    "second": 4,
  },
  Object {
    "first": 6,
    "second": 4,
  },
  Object {
    "first": 6,
    "second": 5,
  },
  Object {
    "first": 6,
    "second": 5,
  },
  Object {
    "first": 6,
    "second": 6,
  },
  Object {
    "first": 6,
    "second": 6,
  },
  Object {
    "first": 6,
    "second": 6,
  },
  Object {
    "first": 6,
    "second": 6,
  },
  Object {
    "first": 7,
    "second": 2,
  },
  Object {
    "first": 7,
    "second": 2,
  },
  Object {
    "first": 7,
    "second": 3,
  },
  Object {
    "first": 7,
    "second": 3,
  },
  Object {
    "first": 7,
    "second": 4,
  },
  Object {
    "first": 7,
    "second": 4,
  },
  Object {
    "first": 7,
    "second": 5,
  },
  Object {
    "first": 7,
    "second": 5,
  },
  Object {
    "first": 7,
    "second": 6,
  },
  Object {
    "first": 7,
    "second": 6,
  },
  Object {
    "first": 7,
    "second": 6,
  },
  Object {
    "first": 7,
    "second": 6,
  },
  Object {
    "first": 7,
    "second": 7,
  },
  Object {
    "first": 7,
    "second": 7,
  },
  Object {
    "first": 7,
    "second": 7,
  },
  Object {
    "first": 7,
    "second": 7,
  },
  Object {
    "first": 8,
    "second": 2,
  },
  Object {
    "first": 8,
    "second": 2,
  },
  Object {
    "first": 8,
    "second": 3,
  },
  Object {
    "first": 8,
    "second": 3,
  },
  Object {
    "first": 8,
    "second": 4,
  },
  Object {
    "first": 8,
    "second": 4,
  },
  Object {
    "first": 8,
    "second": 5,
  },
  Object {
    "first": 8,
    "second": 5,
  },
  Object {
    "first": 8,
    "second": 6,
  },
  Object {
    "first": 8,
    "second": 6,
  },
  Object {
    "first": 8,
    "second": 6,
  },
  Object {
    "first": 8,
    "second": 6,
  },
  Object {
    "first": 8,
    "second": 7,
  },
  Object {
    "first": 8,
    "second": 7,
  },
  Object {
    "first": 8,
    "second": 7,
  },
  Object {
    "first": 8,
    "second": 7,
  },
  Object {
    "first": 8,
    "second": 8,
  },
  Object {
    "first": 8,
    "second": 8,
  },
  Object {
    "first": 8,
    "second": 8,
  },
  Object {
    "first": 8,
    "second": 8,
  },
  Object {
    "first": 9,
    "second": 2,
  },
  Object {
    "first": 9,
    "second": 2,
  },
  Object {
    "first": 9,
    "second": 3,
  },
  Object {
    "first": 9,
    "second": 3,
  },
  Object {
    "first": 9,
    "second": 4,
  },
  Object {
    "first": 9,
    "second": 4,
  },
  Object {
    "first": 9,
    "second": 5,
  },
  Object {
    "first": 9,
    "second": 5,
  },
  Object {
    "first": 9,
    "second": 6,
  },
  Object {
    "first": 9,
    "second": 6,
  },
  Object {
    "first": 9,
    "second": 6,
  },
  Object {
    "first": 9,
    "second": 6,
  },
  Object {
    "first": 9,
    "second": 7,
  },
  Object {
    "first": 9,
    "second": 7,
  },
  Object {
    "first": 9,
    "second": 7,
  },
  Object {
    "first": 9,
    "second": 7,
  },
  Object {
    "first": 9,
    "second": 8,
  },
  Object {
    "first": 9,
    "second": 8,
  },
  Object {
    "first": 9,
    "second": 8,
  },
  Object {
    "first": 9,
    "second": 8,
  },
  Object {
    "first": 9,
    "second": 9,
  },
  Object {
    "first": 9,
    "second": 9,
  },
  Object {
    "first": 9,
    "second": 9,
  },
  Object {
    "first": 9,
    "second": 9,
  },
  Object {
    "first": 10,
    "second": 2,
  },
  Object {
    "first": 10,
    "second": 3,
  },
  Object {
    "first": 10,
    "second": 4,
  },
  Object {
    "first": 10,
    "second": 5,
  },
  Object {
    "first": 10,
    "second": 6,
  },
  Object {
    "first": 10,
    "second": 6,
  },
  Object {
    "first": 10,
    "second": 7,
  },
  Object {
    "first": 10,
    "second": 7,
  },
  Object {
    "first": 10,
    "second": 8,
  },
  Object {
    "first": 10,
    "second": 8,
  },
  Object {
    "first": 10,
    "second": 9,
  },
  Object {
    "first": 10,
    "second": 9,
  },
  Object {
    "first": 10,
    "second": 10,
  },
  Object {
    "first": 11,
    "second": 2,
  },
  Object {
    "first": 11,
    "second": 3,
  },
  Object {
    "first": 11,
    "second": 4,
  },
  Object {
    "first": 11,
    "second": 5,
  },
  Object {
    "first": 11,
    "second": 6,
  },
  Object {
    "first": 11,
    "second": 6,
  },
  Object {
    "first": 11,
    "second": 7,
  },
  Object {
    "first": 11,
    "second": 7,
  },
  Object {
    "first": 11,
    "second": 8,
  },
  Object {
    "first": 11,
    "second": 8,
  },
  Object {
    "first": 11,
    "second": 9,
  },
  Object {
    "first": 11,
    "second": 9,
  },
  Object {
    "first": 11,
    "second": 10,
  },
  Object {
    "first": 11,
    "second": 11,
  },
  Object {
    "first": 12,
    "second": 2,
  },
  Object {
    "first": 12,
    "second": 3,
  },
  Object {
    "first": 12,
    "second": 4,
  },
  Object {
    "first": 12,
    "second": 5,
  },
  Object {
    "first": 12,
    "second": 6,
  },
  Object {
    "first": 12,
    "second": 6,
  },
  Object {
    "first": 12,
    "second": 7,
  },
  Object {
    "first": 12,
    "second": 7,
  },
  Object {
    "first": 12,
    "second": 8,
  },
  Object {
    "first": 12,
    "second": 8,
  },
  Object {
    "first": 12,
    "second": 9,
  },
  Object {
    "first": 12,
    "second": 9,
  },
  Object {
    "first": 12,
    "second": 10,
  },
  Object {
    "first": 12,
    "second": 11,
  },
  Object {
    "first": 12,
    "second": 12,
  },
]
`;
